src/main/java/com/commercetools/sync/products/ProductSync.java

```diff
@@ -181,11 +181,16 @@ private void processFetchedProducts(@Nonnull final Set<Product> fetchedProducts)
     }
 
     @Nonnull
-    private CompletionStage<List<Optional<Product>>> createOrUpdateProducts() {
-        return productService.createProducts(draftsToCreate)
-                             .thenAccept(createdProducts ->
-                                 updateStatistics(createdProducts, draftsToCreate.size()))
-                             .thenCompose(ignoredResult -> syncProducts(productsToSync));
+    private CompletionStage<Void> createOrUpdateProducts() {
+        final CompletableFuture<Void> futureCreates =
+            productService.createProducts(draftsToCreate)
+                          .thenAccept(createdProducts -> updateStatistics(createdProducts, draftsToCreate.size()))
+                          .toCompletableFuture();
+
+        final CompletableFuture<List<Optional<Product>>> futureUpdates =
+            syncProducts(productsToSync).toCompletableFuture();
```

Comment:
`ProductService#createProducts` and `ProductSync#syncProducts` could return more specific `CompletableFuture<T>` instead of `CompletionStage<T>`, then `toCompletableFuture()` becomes redundant!

Actually, i think we did design mistake from the beginning: it is generally recommended to return as narrow (low) class as possible, and for most of our functions we have to return `CompletableFuture`!

Reply:
Indeed, I agree :) for now at least: a60dfcf. For the rest issue will be created ;)

Reply:
https://github.com/commercetools/commercetools-sync-java/issues/271